### PR TITLE
feat(aws-fargate-secretsmanager): Create new construct

### LINKED ISSUE #669: feat(aws-fargate-secretsmanager): new construct
Create a construct that links a Fargate Service to AWS Secrets Manager

Should create a VPC or accept an existing VPC
Should create a Fargate Service or accept an existing service
Should create a new secret object or accept an existing secret
Should grant the service IAM permission to read/write from the secret
Should set up environment variables in the service that identify the secret in the construct code so the container can read/write from the secret

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by mickychetta**
*Issue #669 , if available:*

*Description of changes:*
Create a construct that links a Fargate Service to AWS Secrets Manager

Should create a VPC or accept an existing VPC
Should create a Fargate Service or accept an existing service
Should create a new secret object or accept an existing secret
Should grant the service IAM permission to read/write from the secret
Should set up environment variables in the service that identify the secret in the construct code so the container can read/write from the secret

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

fixes #669

### CODE REVIEW

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: bd63b25f8436a9e550ced0a654d5bcd06a269b0a
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=553c12ea-dd6f-4fee-95bc-f9128a57bafe%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: bd63b25f8436a9e550ced0a654d5bcd06a269b0a
* Result: FAILED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f642a079-7f7f-480e-afa8-646bbab1a2da%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/aws-fargate-secretsmanager/README.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+### AWS Fargate Service
+* Sets up an AWS Fargate service
+  * Uses the existing service if provided
+  * Creates a new service if none provided.
+    * Service will run in isolated subnets if available, then private subnets if available and finally public subnets
+  * Adds environment variables to the container with the ARN and Name of the Secrets Manager secret
+  * Add permissions to the container IAM role allowing it to publish to the Secrets Manager secret
+
+### Amazon Secrets Manager Secret
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/aws-fargate-secretsmanager/README.md:111]: Compare these with aws-lambda-secretmanager

* This should mention random name and random value
* Explore whether the KMS comment adds value - I don't think we have to do anything with KMS when using secret manager. It's possible the right thing to do is remove that comment from the lambda-secretmanager README file.

**Review by biffgaut — APPROVED** [review]
One comment to resolve, but everything substantive is fine and you can start building

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 88d16489ce654ecfd1231a7b4adca25cf5a5230e
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ab35c57b-d320-4170-a12d-fbe87b061bcb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 88d16489ce654ecfd1231a7b4adca25cf5a5230e
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1fbe1a3b-25da-4e4d-8553-b1a20591039b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/aws-fargate-secretsmanager/lib/index.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+   * by this construct.
+   *
+   * @default - none
+   */
+  readonly publicApi: boolean;
+  /**
+   * Optional custom properties for a VPC the construct will create. This VPC will
+   * be used by the new Fargate service the construct creates (that's
+   * why targetGroupProps can't include a VPC). Providing
+   * both this and existingVpc is an error. An Secrets Manager Interface
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/aws-fargate-secretsmanager/lib/index.ts:33]: A Secrets Manager Interface...

(not "an")

File: source/patterns/@aws-solutions-constructs/aws-fargate-secretsmanager/lib/index.ts [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+   * endpoint will be included in this VPC.
+   *
+   * @default - none
+   */
+  readonly vpcProps?: ec2.VpcProps;
+  /**
+   * An existing VPC in which to deploy the construct. Providing both this and
+   * vpcProps is an error. If the client provides an existing Fargate service,
+   * this value must be the VPC where the service is running. An Secrets Manager Interface
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/aws-fargate-secretsmanager/lib/index.ts:42]: A Secrets Manager Interface...

File: source/patterns/@aws-solutions-constructs/aws-fargate-secretsmanager/test/fargate-secretsmanager.test.ts [hunk 4]
```diff
@@ -0,0 +1,639 @@
...
+});
+
+test('New service/new secret, private API, new VPC', () => {
+  const stack = new cdk.Stack();
+  const publicApi = false;
+
+  new FargateToSecretsmanager(stack, 'test-construct', {
+    publicApi,
+    ecrRepositoryArn: defaults.fakeEcrRepoArn,
+    vpcProps: { cidr: '172.0.0.0/16' },
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/aws-fargate-secretsmanager/test/fargate-secretsmanager.test.ts:148]: Third time I see this string hardcoded in the code I need to jump in and say make it a constant.

**Review by biffgaut — CHANGES_REQUESTED** [review]
My comments from the original README review were not addressed (mention that the secret created will have a random name and random value)

File: source/patterns/@aws-solutions-constructs/aws-fargate-secretsmanager/README.md (hunk 1 above)
Comment (mickychetta) [reply source/patterns/@aws-solutions-constructs/aws-fargate-secretsmanager/README.md:111]: KMS comment does not add value

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 20c4a004fd888de19373109411a505ba00e8c19e
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=738c4273-1aaf-443e-85ed-9ce05ec5adcc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 20c4a004fd888de19373109411a505ba00e8c19e
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b1851bf6-5603-4049-bcca-655d68d90dfd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by biffgaut — APPROVED** [review]

**Pull request merged**